<a href="https://colab.research.google.com/github/rw61475-cloud/Tugas-Data-Preprocessing/blob/main/Tugas1/2433027_RiniWidiastuti_TugasBI1_Preprocessing_MissingValue.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

A. Tahap Import Library

1. import numpy as np
   NumPy adalah library Python yang dipakai untuk mengolah data berbentuk array atau matriks dengan lebih cepat dan efisien. as np dipakai supaya kita bisa menyingkat penulisan NumPy menjadi np saat digunakan di kode.
2. import matplotlib.pyplot as plt
   Matplotlib.pyplot adalah library Python yang digunakan untuk membuat grafik atau visualisasi data. as plt dipakai supaya penulisannya lebih singkat saat membuat plot.
3. import pandas as pd
   Pandas adalah library Python yang dipakai untuk membaca, mengolah, dan mengatur data dari file seperti CSV, Excel, atau Text ke dalam bentuk tabel (DataFrame). as pd digunakan supaya penulisannya lebih ringkas saat dipakai di kode.



In [24]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

B. Import Dataset

1. Untuk membaca dataset, kita bisa menggunakan `dataset = pd.read_csv(...)` lalu di dalam tanda kurung diisi dengan lokasi atau path file dataset yang ingin dibuka.
lalu, cara ini harus disesuaikan dengan jenis file yang digunakan. Kalau file berbentuk CSV atau text, kita pakai `pd.read_csv`. Sedangkan kalau file Excel (.xlsx), kita pakai `pd.read_excel`.
2. x = dataset.iloc[:, :-1].values
   Variabel x berisi fitur atau atribut (data input). Bagian ini mengambil semua baris, tapi hanya kolom dari awal sampai sebelum kolom terakhir.
3. y = dataset.iloc[:, -1].values
   Variabel y berisi label atau hasil yang ingin diprediksi. Di sini hanya diambil kolom paling terakhir dari dataset.
4. .iloc
   Digunakan untuk mengambil data berdasarkan posisi indeks (angka), bukan nama kolom. Jadi kita bisa pilih baris atau kolom pakai urutan angkanya.
5. .values
   Digunakan untuk mengambil isi data saja dalam bentuk array NumPy, tanpa nama kolom atau indeksnya.


In [25]:
dataset = pd.read_csv('/content/Dataset_Keputusan_Pembelian_2433027.csv')
x = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

Output print(x) menampilkan isi variabel atribut (x) dalam bentuk array NumPy yang diambil langsung dari file CSV sebelum proses preprocessing.

Matriks ini memiliki 3 kolom utama:

1. Kolom 0 berisi data kategori seperti nama kota (misalnya Jakarta, Bandung, Magelang, Medan, Yogyakarta)
2. Kolom 1 berisi data usia
3. Kolom 2 berisi data pendapatan atau harga

Dari output juga terlihat adanya nilai nan, yang artinya data kosong atau hilang. Ini menunjukkan bahwa dataset masih belum bersih dan perlu dilakukan proses pengisian data (imputasi) sebelum digunakan untuk machine learning.

In [26]:
print (x)

[['Jakarta' 34.0 6700.0]
 ['Bandung' 43.0 nan]
 ['Magelang' 32.0 6700.0]
 [nan 51.0 7100.0]
 ['Yogyakarta' nan 6500.0]
 ['Jakarta' nan 7800.0]
 ['Bandung' 36.0 nan]
 ['Magelang' 49.0 7100.0]
 ['Medan' 32.0 5500.0]
 ['Yogyakarta' nan 7600.0]
 [nan 43.0 5600.0]
 ['Bandung' 30.0 nan]
 ['Magelang' 46.0 7900.0]
 ['Medan' 27.0 8800.0]
 ['Yogyakarta' 29.0 6700.0]
 ['Jakarta' nan 3400.0]
 ['Bandung' nan 7600.0]
 ['Magelang' 38.0 8600.0]
 ['Medan' nan nan]
 ['Yogyakarta' 49.0 6700.0]
 ['Jakarta' 29.0 nan]
 ['Bandung' 26.0 7900.0]
 ['Magelang' nan 7600.0]
 ['Medan' 42.0 6600.0]
 ['Yogyakarta' 25.0 5300.0]
 ['Jakarta' 38.0 nan]
 ['Bandung' 45.0 nan]
 ['Magelang' 43.0 5600.0]
 ['Medan' 50.0 7400.0]
 ['Yogyakarta' 31.0 6200.0]]


Output print(y) menunjukkan isi variabel target (label) dalam bentuk array NumPy yang diambil dari kolom terakhir dataset.

Data ini berisi hasil keputusan pembelian konsumen, yaitu “Ya” dan “Tidak”, yang masih dalam bentuk teks.

In [27]:
print (y)

['Ya' 'Tidak' 'Ya' 'Ya' 'Ya' 'Tidak' 'Ya' 'Ya' 'Tidak' 'Ya' 'Tidak' 'Ya'
 'Ya' 'Tidak' 'Ya' 'Tidak' 'Ya' 'Ya' 'Tidak' 'Ya' 'Tidak' 'Ya' 'Ya'
 'Tidak' 'Ya' 'Tidak' 'Ya' 'Tidak' 'Ya' 'Ya']


C. Menghilangkan Missing Value (nan)

1. Bagaimana kita bisa mengisi data yang kosong secara otomatis? Dengan menggunakan kelas "SimpleImputer" yang diimpor dari library sklearn.impute. Kelas ini mendefinisikan strategi pengisian nilai yang hilang pada dataset.
2. missing_values=np.nan, Digunakan untuk memberi tahu bahwa data kosong di dataset ditandai dengan NaN.
3. strategy='mean', Menentukan cara pengisian data kosong, yaitu menggunakan nilai rata-rata dari kolom tersebut. bisa juga menggunakan median, 'most_frequent' atau modus.
4. imputer.fit(x[:, 1:3]), Digunakan untuk menghitung nilai rata-rata dari kolom ke-1 sampai ke-2 (kolom indeks 1 dan 2), lalu menyimpannya di imputer.
5. imputer.transform(x[:, 1:3]), Digunakan untuk mengganti data yang kosong dengan nilai rata-rata yang sudah dihitung sebelumnya, lalu hasilnya dimasukkan kembali ke dataset.

In [28]:
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
imputer.fit(x[:, 1:3])
x[:, 1:3] = imputer.transform(x[:, 1:3])

hasil print (x) pada tahap ini

Output matriks x setelah proses SimpleImputer menunjukkan bahwa data yang kosong sudah berhasil diisi menggunakan nilai rata-rata (mean).

Pada kolom numerik:
1. Kolom usia (indeks 1) yang sebelumnya kosong sudah diisi dengan nilai rata-rata sekitar 37.73
2. Kolom pendapatan (indeks 2) yang kosong juga sudah diisi dengan nilai rata-rata sekitar 6821.73

Sedangkan nilai NaN pada kolom kota (kolom 0) masih tetap ada, karena proses imputasi hanya dilakukan pada kolom numerik saja ([:, 1:3]). Selain itu, metode mean memang hanya bisa digunakan untuk data angka, bukan data teks seperti nama kota.

Nantinya, kolom kota ini akan ditangani pada tahap encoding (One-Hot Encoding).

In [29]:
print (x)

[['Jakarta' 34.0 6700.0]
 ['Bandung' 43.0 6821.739130434783]
 ['Magelang' 32.0 6700.0]
 [nan 51.0 7100.0]
 ['Yogyakarta' 37.73913043478261 6500.0]
 ['Jakarta' 37.73913043478261 7800.0]
 ['Bandung' 36.0 6821.739130434783]
 ['Magelang' 49.0 7100.0]
 ['Medan' 32.0 5500.0]
 ['Yogyakarta' 37.73913043478261 7600.0]
 [nan 43.0 5600.0]
 ['Bandung' 30.0 6821.739130434783]
 ['Magelang' 46.0 7900.0]
 ['Medan' 27.0 8800.0]
 ['Yogyakarta' 29.0 6700.0]
 ['Jakarta' 37.73913043478261 3400.0]
 ['Bandung' 37.73913043478261 7600.0]
 ['Magelang' 38.0 8600.0]
 ['Medan' 37.73913043478261 6821.739130434783]
 ['Yogyakarta' 49.0 6700.0]
 ['Jakarta' 29.0 6821.739130434783]
 ['Bandung' 26.0 7900.0]
 ['Magelang' 37.73913043478261 7600.0]
 ['Medan' 42.0 6600.0]
 ['Yogyakarta' 25.0 5300.0]
 ['Jakarta' 38.0 6821.739130434783]
 ['Bandung' 45.0 6821.739130434783]
 ['Magelang' 43.0 5600.0]
 ['Medan' 50.0 7400.0]
 ['Yogyakarta' 31.0 6200.0]]


D. Mengubah Data Kategorikal Menjadi Angka (Encoding Data - Atribut)
1. Bagaimana cara mengubah kolom teks atau kategori menjadi bentuk angka biner? Dengan mengimpor kelas "ColumnTransformer" untuk menerapkan transformasi khusus pada kolom tertentu, dan kelas "OneHotEncoder" untuk mengubah data kategori menjadi matriks biner (angka 0 dan 1).
2. transformers=[('encoder', OneHotEncoder(), [0])], Bagian ini mengatur bahwa proses One-Hot Encoding hanya dilakukan pada kolom pertama (kolom indeks 0), sedangkan kolom lain tidak ikut diproses dengan cara ini.
3. remainder='passthrough', Artinya kolom lain yang tidak di-encode tetap dipertahankan apa adanya, tidak dihapus atau diubah.
4. ct.fit_transform(x), Digunakan untuk mempelajari kategori yang ada di data sekaligus langsung mengubahnya menjadi bentuk angka biner (0 dan 1).
5. np.array(...), Digunakan untuk mengubah hasil transformasi menjadi array NumPy supaya lebih rapi dan siap dipakai untuk proses machine learning selanjutnya.

In [30]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer (transformers=[('encoder', OneHotEncoder(), [0])], remainder='passthrough')
x = np.array(ct.fit_transform(x))

Penjelasan dari print (x) pad atahap ini

1. Kolom indeks 0-5 adalah hasil dari One-Hot Encoding. Awalnya 1 kolom kategori dipecah menjadi 6 kolom biner karena ada 6 jenis kategori. Nilai 1.0 artinya data termasuk kategori tersebut, sedangkan 0.0 artinya tidak.
2. Kolom indeks 6 adalah data usia asli yang tetap dipertahankan karena remainder='passthrough'. Di kolom ini juga terlihat nilai desimal karena data kosong sudah diisi dengan nilai rata-rata (mean).
3. Kolom indeks 7 adalah kolom numerik terakhir (misalnya pendapatan/harga) yang juga tidak di-encode. Nilai kosong di kolom ini juga sudah diisi otomatis menggunakan mean, sehingga muncul angka pecahan.


In [31]:
print (x)

[[0.0 1.0 0.0 0.0 0.0 0.0 34.0 6700.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 43.0 6821.739130434783]
 [0.0 0.0 1.0 0.0 0.0 0.0 32.0 6700.0]
 [0.0 0.0 0.0 0.0 0.0 1.0 51.0 7100.0]
 [0.0 0.0 0.0 0.0 1.0 0.0 37.73913043478261 6500.0]
 [0.0 1.0 0.0 0.0 0.0 0.0 37.73913043478261 7800.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 36.0 6821.739130434783]
 [0.0 0.0 1.0 0.0 0.0 0.0 49.0 7100.0]
 [0.0 0.0 0.0 1.0 0.0 0.0 32.0 5500.0]
 [0.0 0.0 0.0 0.0 1.0 0.0 37.73913043478261 7600.0]
 [0.0 0.0 0.0 0.0 0.0 1.0 43.0 5600.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 30.0 6821.739130434783]
 [0.0 0.0 1.0 0.0 0.0 0.0 46.0 7900.0]
 [0.0 0.0 0.0 1.0 0.0 0.0 27.0 8800.0]
 [0.0 0.0 0.0 0.0 1.0 0.0 29.0 6700.0]
 [0.0 1.0 0.0 0.0 0.0 0.0 37.73913043478261 3400.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 37.73913043478261 7600.0]
 [0.0 0.0 1.0 0.0 0.0 0.0 38.0 8600.0]
 [0.0 0.0 0.0 1.0 0.0 0.0 37.73913043478261 6821.739130434783]
 [0.0 0.0 0.0 0.0 1.0 0.0 49.0 6700.0]
 [0.0 1.0 0.0 0.0 0.0 0.0 29.0 6821.739130434783]
 [1.0 0.0 0.0 0.0 0.0 0.0 26.0 7900.0]
 [0.0 0.0

E. Mengubah Data Kategorikal Menjadi Angka (Encoding Data - Label)
1. Bagaimana cara mengubah data teks atau kategori pada variabel label (y) menjadi bentuk numerik sederhana? Kita bisa menggunakan "LabelEncoder". Berbeda dengan OneHotEncoder yang memecah data menjadi beberapa kolom biner (0 dan 1), LabelEncoder mengubah teks menjadi angka urut (seperti 0, 1, 2, dst.) di dalam satu kolom yang sama.
2. le = LabelEncoder(), Digunakan untuk membuat objek LabelEncoder yang akan dipakai untuk mengubah data teks menjadi angka.
3. y = le.fit_transform(y), Digunakan untuk mengubah data label (misalnya “Ya” atau “Tidak”) menjadi bentuk angka (0, 1, dst.) dalam satu kolom yang sama, lalu hasilnya disimpan kembali ke y.

In [32]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
y = le.fit_transform(y)

hasil print (y) pada tahap ini

1. Output array y setelah proses LabelEncoder menunjukkan bahwa data label sudah berubah dari teks menjadi angka.
2. LabelEncoder secara otomatis mengurutkan kategori berdasarkan abjad, lalu mengubahnya menjadi angka. Biasanya, “Tidak” akan menjadi 0 dan “Ya” menjadi 1.
3. Jumlah datanya tetap sama, yaitu 30 elemen, karena hanya bentuknya saja yang berubah, bukan jumlah datanya. Hal ini penting supaya data tetap seimbang dengan dataset awal, tetapi sudah dalam bentuk angka yang bisa diproses oleh machine learning.

In [33]:
print (y)

[1 0 1 1 1 0 1 1 0 1 0 1 1 0 1 0 1 1 0 1 0 1 1 0 1 0 1 0 1 1]


F. Membagi Dataset Menjadi Training Set dan Test Set
1. Bagaimana cara membagi data untuk kebutuhan latihan model dan ujian model? Kita bisa menggunakan fungsi "train_test_split" yang diimpor dari library sklearn.model_selection.
2. x_train dan y_train, Digunakan untuk menyimpan data yang dipakai untuk melatih model machine learning.
3. x_test dan y_test, Digunakan untuk menyimpan data yang dipakai untuk menguji performa model setelah dilatih.
4. test_size = 0.2, Mengatur pembagian data, yaitu 20% untuk data uji dan 80% untuk data latih.
5. random_state = 1, Digunakan supaya hasil pembagian data tetap sama setiap kali program dijalankan.

In [34]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 1)

hasil dari print (x_train) pada tahap ini

Jumlah barisnya jadi 24 karena kita membagi dataset menggunakan  test_size = 0.2. Artinya, dari total 30 data, 20% (6 baris) dipakai sebagai data uji (x_test), sedangkan sisanya 80% (24 baris) dipakai untuk data latih (x_train).

Jadi, hasil x_train berkurang menjadi 24 baris karena sebagian datanya sudah dipisahkan ke data test. Struktur kolomnya tetap sama seperti hasil preprocessing sebelumnya, yaitu 6 kolom hasil One-Hot Encoding, ditambah kolom usia, dan kolom pendapatan yang sudah diisi nilai rata-rata.


In [35]:
print (x_train)

[[1.0 0.0 0.0 0.0 0.0 0.0 45.0 6821.739130434783]
 [0.0 0.0 0.0 0.0 0.0 1.0 51.0 7100.0]
 [0.0 0.0 0.0 0.0 1.0 0.0 25.0 5300.0]
 [0.0 0.0 1.0 0.0 0.0 0.0 37.73913043478261 7600.0]
 [0.0 0.0 0.0 1.0 0.0 0.0 42.0 6600.0]
 [0.0 0.0 0.0 0.0 1.0 0.0 37.73913043478261 6500.0]
 [0.0 0.0 1.0 0.0 0.0 0.0 32.0 6700.0]
 [0.0 1.0 0.0 0.0 0.0 0.0 38.0 6821.739130434783]
 [1.0 0.0 0.0 0.0 0.0 0.0 36.0 6821.739130434783]
 [0.0 0.0 0.0 1.0 0.0 0.0 37.73913043478261 6821.739130434783]
 [0.0 0.0 0.0 1.0 0.0 0.0 27.0 8800.0]
 [0.0 0.0 1.0 0.0 0.0 0.0 49.0 7100.0]
 [0.0 0.0 1.0 0.0 0.0 0.0 43.0 5600.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 43.0 6821.739130434783]
 [1.0 0.0 0.0 0.0 0.0 0.0 37.73913043478261 7600.0]
 [0.0 1.0 0.0 0.0 0.0 0.0 34.0 6700.0]
 [0.0 1.0 0.0 0.0 0.0 0.0 37.73913043478261 3400.0]
 [0.0 0.0 0.0 0.0 1.0 0.0 31.0 6200.0]
 [0.0 0.0 0.0 1.0 0.0 0.0 50.0 7400.0]
 [0.0 0.0 0.0 0.0 1.0 0.0 37.73913043478261 7600.0]
 [0.0 0.0 0.0 1.0 0.0 0.0 32.0 5500.0]
 [0.0 0.0 1.0 0.0 0.0 0.0 46.0 7900.0]
 [1.0 0.0

hasil print (x_test) pada tahap ini

1. Output matriks x_test adalah data atribut yang dipakai untuk menguji model setelah proses training. Data ini sengaja dipisahkan dari data latih supaya bisa dipakai sebagai bahan evaluasi yang benar-benar baru untuk melihat performa model.

2. Jumlah barisnya hanya 6 karena kita memakai test_size = 0.2, artinya 20% dari total 30 data digunakan sebagai data uji. Jadi 20% × 30 = 6 baris, sedangkan 24 sisanya menjadi data latih (x_train).

3. Struktur kolomnya sama seperti data latih, yaitu 6 kolom pertama hasil One-Hot Encoding (berisi 0.0 dan 1.0), lalu kolom usia di kolom ke-7, dan kolom pendapatan di kolom ke-8. Nilai seperti 6821.739130434783 menunjukkan bahwa data yang sebelumnya kosong sudah diisi dengan nilai rata-rata (mean).

In [36]:
print (x_test)

[[0.0 0.0 1.0 0.0 0.0 0.0 38.0 8600.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 26.0 7900.0]
 [0.0 0.0 0.0 0.0 0.0 1.0 43.0 5600.0]
 [0.0 0.0 0.0 0.0 1.0 0.0 49.0 6700.0]
 [0.0 0.0 0.0 0.0 1.0 0.0 29.0 6700.0]
 [0.0 1.0 0.0 0.0 0.0 0.0 29.0 6821.739130434783]]


hasil dari print (y_train) pada tahap ini

1. Output y_train adalah data label (target) yang digunakan untuk melatih model machine learning. Data ini berpasangan langsung dengan x_train, sehingga setiap nilai di y_train sesuai dengan baris data pada x_train.
2. Isinya hanya 0 dan 1 karena sebelumnya kita sudah menggunakan LabelEncoder(), yang mengubah label kategori seperti “Tidak” dan “Ya” menjadi angka. Biasanya “Tidak” jadi 0 dan “Ya” jadi 1.
3. Jumlah datanya ada 24 karena sesuai pembagian data sebelumnya, 80% dari 30 data digunakan untuk training, jadi 24 data masuk ke y_train dan sisanya 6 masuk ke y_test.

In [37]:
print (y_train)

[1 1 1 1 0 1 1 0 1 0 0 1 0 0 1 1 0 1 1 1 0 1 1 0]


hasil print (y_test) pada tahap ini

1. Output y_test adalah data label untuk data uji yang nanti dipakai untuk mengecek hasil prediksi model. Data ini tidak digunakan saat proses training, karena fungsinya sebagai “jawaban asli” untuk membandingkan apakah prediksi model sudah benar atau belum.
2. Jumlah nilainya ada 6 karena sesuai pembagian sebelumnya (test_size = 0.2), yaitu 20% dari 30 data dipakai sebagai data uji. Jadi 6 nilai ini berpasangan langsung dengan 6 baris pada x_test.
3. Angka 0 dan 1 berasal dari proses LabelEncoder(), yang mengubah label teks seperti “Tidak” dan “Ya” menjadi bentuk angka, biasanya 0 untuk kategori pertama dan 1 untuk kategori kedua.

In [38]:
print (y_test)

[1 1 0 1 1 0]


G. Feature Scaling (Standardisasi Ukuran Data)

Feature scaling diperlukan supaya semua fitur numerik berada dalam skala yang seimbang. Kalau tidak, kolom dengan angka besar (misalnya pendapatan) bisa lebih dominan dibanding kolom dengan angka kecil (misalnya usia), sehingga hasil model jadi tidak akurat.
1. from sklearn.preprocessing import StandardScaler, Digunakan untuk memanggil StandardScaler, yaitu alat untuk menstandarkan data agar memiliki rata-rata 0 dan standar deviasi 1.
2. sc = StandardScaler(), Digunakan untuk membuat objek scaler yang nantinya dipakai untuk proses scaling data.
3. x_train[:, 3:] = sc.fit_transform(x_train[:, 3:]), Digunakan untuk menghitung sekaligus mengubah skala data training pada kolom numerik (mulai dari kolom ke-3 sampai terakhir). Bagian ini hanya dilakukan pada data training agar model belajar dari data tersebut.
4. x_test[:, 3:] = sc.transform(x_test[:, 3:]), Digunakan untuk mengubah skala data test menggunakan aturan yang sudah dipelajari dari data training, tanpa menghitung ulang.

In [39]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train[:, 3:] = sc.fit_transform(x_train[:, 3:])
x_test[:, 3:] = sc.transform(x_test[:, 3:])

Hasil dari print (x_train) pada tahap ini

1. Hasil print(x_train) setelah feature scaling menunjukkan data latih yang sudah dinormalisasi. Nilai yang sebelumnya besar seperti pendapatan atau usia sekarang berubah jadi angka kecil dalam rentang sekitar -3 sampai +3.

2. Kolom 0 sampai 2 tetap berisi 0.0 dan 1.0 karena itu hasil One-Hot Encoding, jadi tidak ikut di-scaling. Sedangkan kolom mulai dari indeks 3 sampai akhir berubah menjadi nilai desimal karena bagian ini adalah data numerik yang diproses oleh StandardScaler.

3. Angka negatif berarti nilai data tersebut di bawah rata-rata, sedangkan angka positif berarti di atas rata-rata. Kalau nilainya mendekati 0, artinya data tersebut hampir sama dengan nilai rata-rata kolomnya.

In [40]:
print (x_train)

[[1.0 0.0 0.0 -0.5129891760425772 -0.447213595499958 -0.2085144140570746
  0.992866922055732 0.055599183819007386]
 [0.0 0.0 0.0 -0.5129891760425772 -0.447213595499958 4.7958315233127164
  1.8763637640354707 0.3224752661502409]
 [0.0 0.0 0.0 -0.5129891760425772 2.2360679774997902 -0.2085144140570746
  -1.952122551210064 -1.4038793914299283]
 [0.0 0.0 1.0 -0.5129891760425772 -0.447213595499958 -0.2085144140570746
  -0.07629229976032857 0.8020182265891768]
 [0.0 0.0 0.0 1.9493588689617931 -0.447213595499958 -0.2085144140570746
  0.5511185010658626 -0.15706769428869496]
 [0.0 0.0 0.0 -0.5129891760425772 2.2360679774997902 -0.2085144140570746
  -0.07629229976032857 -0.2529762863764821]
 [0.0 0.0 1.0 -0.5129891760425772 -0.447213595499958 -0.2085144140570746
  -0.9213762355670353 -0.061159102200907776]
 [0.0 1.0 0.0 -0.5129891760425772 -0.447213595499958 -0.2085144140570746
  -0.03787939358729655 0.055599183819007386]
 [1.0 0.0 0.0 -0.5129891760425772 -0.447213595499958 -0.2085144140570746


hasil dari print (x_test) pada tahap ini

1. Hasil print(x_test) setelah feature scaling menunjukkan data uji yang sudah disesuaikan dengan skala yang sama seperti data latih.

2. Kolom 0 sampai 2 tetap berisi 0.0 dan 1.0 karena itu hasil One-Hot Encoding, jadi tidak ikut diproses oleh StandardScaler. Sedangkan kolom mulai dari indeks 3 sampai akhir berubah menjadi angka desimal karena bagian ini adalah data numerik yang sudah distandardisasi.

3. Data uji harus memakai sc.transform() (bukan fit_transform) supaya skalanya mengikuti data latih, bukan menghitung ulang sendiri. Hal ini penting agar hasil pengujian tetap konsisten dan adil. Nilai seperti 0.0555 menunjukkan bahwa data tersebut sudah disesuaikan dengan rata-rata dari data latih.

In [41]:
print (x_test)

[[0.0 0.0 1.0 -0.5129891760425772 -0.447213595499958 -0.2085144140570746
  -0.03787939358729655 1.7611041474670486]
 [1.0 0.0 0.0 -0.5129891760425772 -0.447213595499958 -0.2085144140570746
  -1.804873077546774 1.0897440028525383]
 [0.0 0.0 0.0 -0.5129891760425772 -0.447213595499958 4.7958315233127164
  0.6983679747291525 -1.1161536151665667]
 [0.0 0.0 0.0 -0.5129891760425772 2.2360679774997902 -0.2085144140570746
  1.5818648167088911 -0.061159102200907776]
 [0.0 0.0 0.0 -0.5129891760425772 2.2360679774997902 -0.2085144140570746
  -1.3631246565569046 -0.061159102200907776]
 [0.0 1.0 0.0 -0.5129891760425772 -0.447213595499958 -0.2085144140570746
  -1.3631246565569046 0.055599183819007386]]


KESIMPULAN

Hasil dari tahapan preprocessing pada dataset yang diuji
1. Missing value sudah ditangani dengan mean untuk data numerik, dan encoding untuk data kategori. Semua data juga sudah berubah menjadi bentuk numerik, yaitu One-Hot Encoding untuk fitur (x) dan LabelEncoder untuk target (y).
2. Data kemudian dibagi menjadi 80% data latih (24 data) dan 20% data uji (6 data). Selain itu, data numerik sudah distandarisasi menggunakan StandardScaler agar semua fitur berada dalam skala yang seimbang.

Hasil akhirnya, dataset sudah bersih, rapi, dan siap digunakan untuk proses pemodelan machine learning.